File: problems_list/剑指OfferII116.省份数量/剑指OfferII116-省份数量.cpp
class Solution {
public:
    unordered_map<int, vector<int>> mp;
    vector<int> province;
    int find(int x){
        if(x!=province[x]) province[x]=find(province[x]);
        return province[x];
    }

    
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n=isConnected.size();
        province.resize(n);
        for(int i=0; i<n; i++){
            province[i]=i;
            for(int j=0; j<i; j++){
                if(isConnected[i][j]!=0){
                    mp[i].push_back(j);
                }
            }
        }

        for(int i=0; i<n; i++){
            for(int j=0; j<mp[i].size(); j++){
                province[find(mp[i][j])]=find(i);
            }
        }

        int ans=0;
        for(int i=0; i<n; i++){
            if(province[i]==i){
                ans++;
            }
        }
        return ans;
    }
};

File: problems_list/剑指OfferII116.省份数量/剑指OfferII116-省份数量.js
var findCircleNum = function(isConnected) {
    const cities = isConnected.length;
    const parent = new Array(cities).fill(0)
                                       .map((element, index) => index);

    for (let i = 0; i < cities; i++) {
        for (let j = i + 1; j < cities; j++) {
            if (isConnected[i][j] == 1) {
                union(parent, i, j);
            }
        }
    }
    let provinces = 0;
    parent.forEach((element, index) => {
        if (element === index) {
            provinces++;
        }
    });

    return provinces;
};

const union = (parent, index1, index2) => {
    parent[find(parent, index1)] = find(parent, index2);
}

const find = (parent, index) => {
    if (parent[index] !== index) {
        parent[index] = find(parent, parent[index]);
    }
    return parent[index];
}